contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java

```diff
@@ -0,0 +1,398 @@
+    }
+
+    // Evaluate formulae
+    if (readerConfig.evaluateFormulae) {
+      this.evaluator = workbook.getCreationHelper().createFormulaEvaluator();
+    }
+    this.workbook.setMissingCellPolicy(Row.MissingCellPolicy.CREATE_NULL_AS_BLANK);
+    this.sheet = getSheet();
+  }
+
+  /**
+   * This helper function gets the column headers.
+   * If the user specified no columns headers it populates an array with field names of field_n
+   */
+  private TupleMetadata defineSchema() {
+    // TODO Use case if there are no column headers
+    SchemaBuilder builder = new SchemaBuilder();
+    return getColumnHeaders(builder);
+  }
+
+  protected TupleMetadata getColumnHeaders(SchemaBuilder builder) {
+    //Get the field names
+    int columnCount = 0;
+    if (readerConfig.headerRow >= 0) {
+      columnCount = sheet.getRow(readerConfig.headerRow).getPhysicalNumberOfCells();
+    } else {
+      columnCount = sheet.getRow(0).getPhysicalNumberOfCells();
+    }
+    this.excelFieldNames = new ArrayList<>(columnCount);
+    this.rowIterator = sheet.iterator();
+
+    //If there are no headers, create columns names of field_n
+    if (readerConfig.headerRow == -1) {
+      String missingFieldName;
+      for (int i = 0; i < columnCount; i++) {
+        missingFieldName = MISSING_FIELD_NAME_HEADER + (i + 1);
+        ExcelFormatPlugin.makeColumn(builder, missingFieldName, TypeProtos.MinorType.VARCHAR);
+        excelFieldNames.add(i, missingFieldName);
+      }
+      return builder.buildSchema();
+    } else if (rowIterator.hasNext()) {
+      //Find the header row
+      while (this.lineCount < readerConfig.headerRow) {
+        Row row = rowIterator.next();
+        this.lineCount++;
+      }
+      //Get the header row and column count
+      Row row = rowIterator.next();
+      this.totalColumnCount = row.getLastCellNum();
+
+      //Read the header row
+      Iterator<Cell> cellIterator = row.cellIterator();
+      int colPosition = 0;
+      String tempColumnName = "";
+
+      while (cellIterator.hasNext()) {
+        Cell cell = cellIterator.next();
+        // TODO Potential NPE if cell evaluation is turned off
+        CellValue cellValue = evaluator.evaluate(cell);
+        switch (cellValue.getCellTypeEnum()) {
+          case STRING:
+            tempColumnName = cell.getStringCellValue()
+              //.replaceAll("_", "__")  // Not sure if we need this
+              .replace(PARSER_WILDCARD, SAFE_WILDCARD).replaceAll("\\.", SAFE_SEPARATOR);
+            ExcelFormatPlugin.makeColumn(builder, tempColumnName, TypeProtos.MinorType.VARCHAR);
+            excelFieldNames.add(colPosition, tempColumnName);
+            break;
+          case NUMERIC:
+            tempColumnName = String.valueOf(cell.getNumericCellValue());
+            ExcelFormatPlugin.makeColumn(builder, tempColumnName, TypeProtos.MinorType.FLOAT8);
+            excelFieldNames.add(colPosition, tempColumnName);
```

Comment:
The bove two lines appear the same, save type, in both cases. Pull out as a function?